File: leetcode/140. Word Break II.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        dp = [False for _ in range(len(s)+1)]
        dp[0] = True
        
        word_set = set(wordDict)
        
        maxlen = max([len(word) for word in word_set])
        
        idx_to_words = defaultdict(list)
        
        i = 0
        while i < len(s):
            if not dp[i]:
                i += 1
                continue
            j = i
            while j < min(len(s), i+maxlen):
                word = s[i:j+1]
                if word in word_set:
                    dp[j+1] = True
                    idx_to_words[j+1].append(word)
                j += 1
            i += 1

        res = []
        
        def get_words(idx, curr):
            if idx == 0:
                res.append(curr)
                return

            for word in idx_to_words[idx]:
                new_s = word if curr == '' else word + ' ' + curr
                get_words(idx-len(word), new_s)
                
        get_words(len(s), '')
        return res
        

File: leetcode/140. Word Break II.kt
class Solution {
    fun wordBreak(s: String, wordDict: List<String>): List<String> {
        val wordSet = wordDict.toSet()
        val idx_to_back = mutableMapOf<Int, MutableList<Int>>()
        val dp = BooleanArray(s.length+1)
        dp[0] = true

        val lens = wordDict.map { it.length }
        val min_len = lens.min()
        val max_len = lens.max()

        for (i in 0..s.length-1) {
            if (!dp[i]) continue
            for (j in i+min_len..i+max_len) {
                if (j > s.length) break
                val sub = s.substring(i, j)
                if (sub in wordSet) {
                    dp[j] = true
                    if (j-1 !in idx_to_back) {
                        idx_to_back[j-1] = mutableListOf<Int>()
                    }
                    idx_to_back[j-1]?.add(i)
                }
            }
        }

        val res = mutableListOf<String>()
        fun find_all(end: Int, curr: MutableList<String>) {
            if (end == -1) {
                res.add(curr.asReversed().joinToString(" "))
                return
            }
            for (start in idx_to_back[end] ?: emptyList()) {
                curr.add(s.substring(start, end+1))
                find_all(start-1, curr)
                curr.removeLast()
            }
        }

        find_all(s.length-1, mutableListOf<String>())
        return res

    }
}
